### PR TITLE
Argument type checking on def

### LINKED ISSUE #8005: unexpected type mismatch: expected String, found Record([])
### Describe the bug

```py
def provideStr [r: record] {
  "test"
} 

def needStr [str: string] {
  echo $str
}

def run [r: record] {
  needStr (provideStr $r)
}

run {a:b}
```
This script errors with:
```
Error: nu::parser::type_mismatch (link)

  × Type mismatch.
    ╭─[/d/dev/nu/test.nu:9:1]
  9 │ def run [r: record] {
 10 │   needStr (provideStr $r)
    ·           ───────┬───────
    ·                  ╰── expected String, found Record([])
 11 │ }
    ╰────
```

But `provideStr` does return a string. Putting the expression around quotes (`(provideStr $r)` -> `$"(provideStr $r)"`) solves this issue but it looks unnecessary.  
The more curious case is, that `r` has to be a parameter and not a variable. This script runs just fine:
```py
def provideStr [r: record] {
  "test"
} 

def needStr [str: string] {
  echo $str
}

let r = {a:b}
needStr (provideStr $r)
```

### How to reproduce

1. create a file named `test.nu` with the content:
```py
def provideStr [r: record] {
  "test"
} 

def needStr [str: string] {
  echo $str
}

def run [r: record] {
  needStr (provideStr $r)
}

run {a:b}
```
2. run it with `nu test.nu`

### Expected behavior

I expect nu to use the output from `provideStr` as positional parameter for the `needStr` command. 

### Screenshots

_No response_

### Configuration

| key                | value                                           |
| ------------------ | ----------------------------------------------- |
| version            | 0.75.0                                          |
| branch             |                                                 |
| commit_hash        |                                                 |
| build_os           | linux-x86_64                                    |
| build_target       | x86_64-unknown-linux-gnu                        |
| rust_version       | rustc 1.68.0-nightly (0442fbabe 2023-01-10)     |
| rust_channel       | nightly-x86_64-unknown-linux-gnu                |
| cargo_version      | cargo 1.68.0-nightly (8c460b223 2023-01-04)     |
| pkg_version        | 0.75.0                                          |
| build_time         | 2023-02-08 17:29:13 +01:00                      |
| build_rust_channel | release                                         |
| features           | database, dataframe, default, trash, which, zip |
| installed_plugins  |                                                 |

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Mehrbod2002**
Fixes #8005 .
Trying to detect function as argument and gain output to avoid error on defining function.

### CODE REVIEW

File: crates/nu-parser/src/parser.rs [hunk 1]
```diff
@@ -950,6 +950,9 @@ pub fn parse_internal_call(
 
         // Parse a positional arg if there is one
         if let Some(positional) = signature.get_positional(positional_idx) {
+            if positional.name == "def_name" {
```
Comment (jntrnr) [inline crates/nu-parser/src/parser.rs:953]: I don't know if we want to make this name magical. What if the user creates a parameter with the name `def_name`?

Comment (Mehrbod2002) [reply crates/nu-parser/src/parser.rs]:
``` python
def provideStr [def_name: record] {
  "def_name_should_print"
}

def needStr [def_name: string] {
  echo $def_name
}

def run [def_name: record] {
  needStr (provideStr $def_name)
}

run {a:b}
```

I tested this code , it throws an error as the real `def_name` shape is String and will detect as mismatch but as the variables are define as `var_id` , `def_name` have None var_id and we can use it to detect which is the real `def_name` , last commit , this example will work

**Comment by Mehrbod2002** [comment]
Ping @sholderbach

File: crates/nu-parser/src/parser.rs (hunk 1 above)
Comment (Mehrbod2002) [reply crates/nu-parser/src/parser.rs]: Consider to review it please or even any other thing can close it .
Thanks @jntrnr

**Comment by Mehrbod2002** [comment]
Consider to take look at this PR.
Thanks a lot .

**Comment by Mehrbod2002** [comment]
Consider to take look at this PR , means lot to me
Thanks a lot .
@fdncred @sholderbach

File: crates/nu-parser/src/parser.rs [hunk 2]
```diff
@@ -950,6 +950,9 @@ pub fn parse_internal_call(
 
         // Parse a positional arg if there is one
         if let Some(positional) = signature.get_positional(positional_idx) {
+            if positional.name == "def_name" && positional.var_id.is_none() {
```
Comment (jntrnr) [inline crates/nu-parser/src/parser.rs:953]: please do not check for a magic parameter name. This is considered bad coding practice.

Comment (Mehrbod2002) [reply crates/nu-parser/src/parser.rs:953]: Is there any way to detect it's parsing def ? Ever we add some new parameter on parse_def function. 
Is there any way ?

Comment (Mehrbod2002) [reply crates/nu-parser/src/parser.rs]: we can check it only with var_id too.
new tests added too .

**Comment by fdncred** [comment]
Making this a draft. Not sure if progress needs to continue on this or not.

**Comment by Mehrbod2002** [comment]
> Making this a draft. Not sure if progress needs to continue on this or not.

It is not process , but branch has a conflict 
Is this problem already solved ? 
If not,  can you consider to checkout a solution and react to it ? Then I can solve its conflicting

**Comment by fdncred** [comment]
@jntrnr can you check to make sure your points have been addressed?

**Comment by fdncred** [comment]
What's the status on this PR?

**Comment by Mehrbod2002** [comment]
> What's the status on this PR?

Will close it ,
No one won't attention to it

**Comment by jntrnr** [comment]
We chatted about this and would rather work on a re-implentation of the type checker first rather than adding more workarounds on what we have currently. Thanks for the work, it'll get us planning something better than what we have today.

**Pull request closed without merging**